statsmodels/stats/robustros.py

```diff
@@ -0,0 +1,477 @@
+            cohn['ncen_equal'] = cohn.apply(ncen_equal, axis=1)
+
+            # add an extra row
+            cohn = cohn.reindex(range(cohn.shape[0]+1))
+
+            # add the 'prob_exceedance' column, initialize with zeros
+            cohn['prob_exceedance'] = 0.0
+
+        else:
+            dl_cols = ['DL', 'lower', 'upper', 'nuncen_above',
+                       'nobs_below', 'ncen_equal', 'prob_exceedance']
+            cohn = pd.DataFrame(np.empty((0,7)), columns=dl_cols)
+
+        return cohn
+
+    def estimate(self):
+        '''
+        Estimates the values of the censored data
+        '''
+        def _ros_DL_index(row):
+            '''
+            Helper function to create an array of indices for the
+            detection  limits (self.cohn) corresponding to each
+            data point
+            '''
+            DLIndex = np.zeros(len(self.data[self.result]))
+            if self.cohn.shape[0] > 0:
+                index, = np.where(self.cohn['DL'] <= row[self.result])
+                DLIndex = index[-1]
+            else:
+                DLIndex = 0
+
+            return DLIndex
+
+        def _ros_plotting_pos(row):
+            '''
+            Helper function to compute the ROS'd plotting position
+            '''
+            dl_1 = self.cohn.iloc[row['DLIndex']]
+            dl_2 = self.cohn.iloc[row['DLIndex']+1]
+            if row[self.censorship]:
+                return (1 - dl_1['prob_exceedance']) * row['Rank']/(dl_1['ncen_equal']+1)
+            else:
+                return (1 - dl_1['prob_exceedance']) + (dl_1['prob_exceedance'] - dl_2['prob_exceedance']) * \
+                        row['Rank'] / (dl_1['nuncen_above']+1)
+
+        def _select_modeled(row):
+            '''
+            Helper fucntion to select "final" data from original detects
+            and estimated non-detects
+            '''
+            if row[self.censorship]:
+                return row['modeled_data']
+            else:
+                return row[self.result]
+
+        def _select_half_DL(row):
+            '''
+            Helper function to select half cohn when there are
+            too few detections
+            '''
+            if row[self.censorship]:
+                return 0.5 * row[self.result]
+            else:
+                return row[self.result]
+
+        # create a DLIndex column that references self.cohn
+        self.data['DLIndex'] = self.data.apply(_ros_DL_index, axis=1)
+
+        # compute the ranks of the data
+        self.data['Rank'] = 1
+        groupcols = ['DLIndex', self.censorship, 'Rank']
+        rankgroups = (
+            self.data[groupcols]
+                .groupby(by=['DLIndex', self.censorship])
+        )
+        self.data['Rank'] = rankgroups.transform(lambda x: x.cumsum())
+
+        # detect/non-detect selectors
+        detect_selector = self.data[self.censorship] == False
+        nondet_selector = self.data[self.censorship] == True
```

Comment:
It's more standard (for me) to call this `mask`  `detect_mask` `nondet_mask`
we might want to switch to censored uncensored terminology instead of nondetect detect, but it's mainly used internally.